### ISSUE TITLE
BUG: scipy.special.stdtri: calling with small `df` crashes kernel

### ISSUE THREAD

**Issue by Gabriel-Kissin** [scipy/scipy#18906]
### Describe your issue.

For certain distributions in scipy.stats, for very low values of parameters, the kernel crashes.  

With other distributions, such low values cause a warning or an exception, which is fine. But with some distributions it simply crashes. It should instead be handled correctly.

Note that all parameters being used are within the domain for that parameter, as shown in the code below. 

I originally found this issue with the beta distribution. Before posting I wrote code which looked through all the other distributions to see if it could create a crash with them also. It only managed with the t distribution. 

With all other distributions, funny values gives 
- a warning like `RuntimeWarning: overflow encountered in [divide/double_scalars/_ncf_ppf/_nct_ppf]`,  `RuntimeWarning: invalid value encountered in [multiply/divide/add/subtract/double_scalars]`, `RuntimeWarning: divide by zero encountered in [divide/log]`, `RuntimeWarning: floating point number truncated to an integer`
- or an error like `RuntimeError: Failed to converge after 100 iterations.` (for `scipy.stats.genexpon(5e-324, 5e-324, 5e-324).ppf(.2)` and `scipy.stats.studentized_range(2, 5e-10).ppf(.2)`). 

### Reproducing Code Example

```python
import scipy.stats
print([shape_param.domain for shape_param in scipy.stats.t._shape_info()])
# [[5e-324, inf]]
print(scipy.stats.t(1e-10).ppf(.2))
# -1e+100
print(scipy.stats.t(1e-20).ppf(.2))
# 1e+100
print(scipy.stats.t(1e-30).ppf(.2))
# KERNEL CRASHES with message: `STOP  SMALL, X, BIG not monotone in INVR`


import scipy.stats
print([shape_param.domain for shape_param in scipy.stats.beta._shape_info()])
# [[5e-324, inf], [5e-324, inf]]
print(scipy.stats.beta(a=1e-308, b=5).ppf(.2))
# 0.0
print(scipy.stats.beta(a=1e-309, b=5).ppf(.2))
# OverflowError: Error in function boost::math::tgamma<d>(d): Overflow Error
# (raises error, but kernel doesn't crash)
print(scipy.stats.beta(a=1e-323, b=5).ppf(.2))
# KERNEL CRASHES with message: `Assertion failed: *p_derivative >= 0, file ..\..\scipy\_lib\boost/boost/math/special_functions/beta.hpp, line 739`
```

### Error message

```shell
# after print(scipy.stats.t(1e-30).ppf(.2))
STOP  SMALL, X, BIG not monotone in INVR

# after print(scipy.stats.beta(a=1e-323, b=5).ppf(.2))
Assertion failed: *p_derivative >= 0, file ..\..\scipy\_lib\boost/boost/math/special_functions/beta.hpp, line 739
```

### SciPy/NumPy/Python version and system information

```shell
1.10.1 1.23.1 sys.version_info(major=3, minor=9, micro=4, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: c:/opt/openblas/if_32/64/include
    lib directory: c:/opt/openblas/if_32/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= PRESCOTT MAX_THREADS=4
    pc file directory: c:/opt/openblas/if_32/64/lib/pkgconfig
    version: 0.3.18
  lapack:
    detection method: pkgconfig
    found: true
    include directory: c:/opt/openblas/if_32/64/include
    lib directory: c:/opt/openblas/if_32/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= PRESCOTT MAX_THREADS=4
    pc file directory: c:/opt/openblas/if_32/64/lib/pkgconfig
    version: 0.3.18
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 0.29.33
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  pythran:
    include directory: C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-u63ta2f1\overlay\Lib\site-packages\pythran
    version: 0.12.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
Python Information:
  path: C:\Users\runneradmin\AppData\Local\Temp\cibw-run-a1px0t3e\cp39-win_amd64\build\venv\Scripts\python.exe
  version: '3.9'
```

**Comment by mdhaber** [comment]
@mborland thought you might be interested in this one, since `beta` is coming from Boost. 
Linking `scipy.special` because that's where `t` is coming from.

**Comment by mdhaber** [comment]
> general crashes should be avoided...

Sure, but how they are to be avoided depends on whether there is a reasonable use case for the returned values. The easy fix is to change the documented domain of the distribution parameters. A harder fix would be to branch with an asymptotic approximation.

**Comment by Gabriel-Kissin** [comment]
Right.

> The easy fix is to change the documented domain of the distribution parameters, so these cases would return NaN. 

That would certainly be much easier, but not sure about that... The parameters' documented domain always matches the mathematical definition of the distribution. For example, the upper limit for beta's `a` and `b` is infinity (`print([shape_param.domain for shape_param in scipy.stats.beta._shape_info()])`), even though largeish numbers fail to give meaningful results. For example, `print(scipy.stats.beta(10e25,2).ppf(.2))` gives a `nan` - but the domain doesn't restrict the parameter to `10e20` because of that. With all the other distributions, the parameters' given domain always matches the distribution's mathematical definition, not the practical domain where the code can give you solid results. So unless that was changed across the board, probably the way to go would be to raise some sort of error or warning. Or an approximation as per your second suggestion would be even better

**Comment by tupui** [comment]
I think Matt meant to say "domain" as in what we allow. i.e. input validation. In practice, we have to deal with numerical accuracy, implementation complexity, time to solution, etc. Moving forward we are thinking about being clearer about what we guarantee and ways to go around if you really want to (e.g. using mpmath.)

**Comment by Gabriel-Kissin** [comment]
Oh right. I knew that input validation is done using the distribution's `._argcheck()` method, but I'd assumed that that relies partially on the distribution's `._shape_info()` and shape_info's `.domain` - so thought they were one and the same. 

Looking at the code for `.argcheck()` using `import inspect ; print(inspect.getsource(scipy.stats.beta._argcheck))`, I see that I was mistaken, and `.argcheck()` stores permissible domains for parameters separately. For both `t` and `beta`, all it requires is that `asarray(arg) > 0`. This `.argcheck()` is inherited from the base class for continuous RVs.

In which case yes, I guess this could be easily patched by working out the exact values which cause problems, and overriding the inherited `.argcheck()` for `beta` and `t` to catch them.

**Comment by tupui** [comment]
Yeah so to be frank patching the current infrastructure is tedious. We have a plan in motion to have a new architecture which would hopefully be more sane. We still find bugs/issues in the current infra after 10+ years 😓

**Comment by mdhaber** [comment]
> In which case yes, I guess this could be easily patched by working out the exact values which cause problems,

Yes, whatever we do, we'll need to understand when these things happen. Is it only the value of the shape(s) that matter, or does the argument of `ppf` play a role? Can you study this @Gabriel-Kissin and, assuming the dimensionality is sufficiently low, present the findings graphically (e.g. like https://github.com/scipy/scipy/pull/18822#issuecomment-1624801367)?

**Comment by Gabriel-Kissin** [comment]
Yes, that should be doable. I was thinking however that as both of these errors originate in the C++ code that that might be the best place to easily tackle them. For example, with beta it appears to be an assertion which is broken in these cases which causes the crash. In the C++ code, the assertion could be handled much better - a warning could be given explaining whatever the reason of the assertion is, a nan could be returned, rather than raising an error which causes crashing in python. That approach might mean that detailed diagnostics into what values cause the issue aren't necessary.  

The other question in my mind as I was planning how this could be done is that since it crashes the kernel, tracking this would be tricky. This error can't be handled the normal way with `try except`, and I'm not aware of how this can be handled from within python. Within C++ this investigation would be easier.  

[The C++ source code which raises the error is at [boost.org](https://www.boost.org/doc/libs/1_53_0/boost/math/special_functions/beta.hpp), one of the lines which say `BOOST_ASSERT(*p_derivative >= 0)` (as in the beta error message in the original post). On [github](https://github.com/boostorg/math/blob/develop/include/boost/math/special_functions/beta.hpp) the same lines say `BOOST_MATH_ASSERT(*p_derivative >= 0)` instead for some reason.]   

Unfortunately I've not used C++ for far too long. Maybe the best way forward would be to open an issue at https://github.com/boostorg/math/issues linking to this discussion?

**Comment by mdhaber** [comment]
I pinged a Boost maintainer above. They'll let us know whether they'd like us to open an issue.

**Comment by mdhaber** [comment]
I think resolving the `beta` part of this issue will be enough to remove the `scipy.stats` label from this one; the `t` part is in `special`.

```python3
import scipy
scipy.special.stdtrit(1e-100, 0.5)
# STOP  SMALL, X, BIG not monotone in INVR
# interpreter crashes
```

**Comment by Gabriel-Kissin** [comment]
> the `t` part is in `special`.

Right. And there is anyway [talk](https://github.com/scipy/scipy/issues/18678) about rewriting `scipy.special.stdtrit`... or making it use boost instead of the current code (which is apparently written in Fortran 77!)

**Comment by mdhaber** [comment]
@Gabriel-Kissin are you able to reproduce the `beta` part of the issue with Boost alone?

**Comment by Gabriel-Kissin** [comment]
Just thought worth adding to this thread - I was using the tensorflow probability module as I needed autodifferentiation functionality. `tfp`'s API is very similar to `scipy`'s, but it has a few added functionalities (e.g. it allows declaring a distribution while giving it a set of different parameterisations, supports GPUs, etc). While using `tfp` I noticed that it doesn't suffer from these issues:  

```
import tensorflow_probability as tfp
print(tfp.distributions.Beta(concentration1=1e-323, concentration0=5).quantile(.2))
# tf.Tensor(nan, shape=(), dtype=float32)
print(tfp.distributions.StudentT(df=1e-30, loc=0, scale=1).quantile(.2))
# tf.Tensor(-inf, shape=(), dtype=float32)
```

Thought this might be helpful to anyone else using silly values in their distributions... plus a possible resolution could be to implement Beta and T the same way as tensorflow does - I'm guessing either tfp uses something other than Boost, or it restricts the allowed params similar to what @mdhaber suggested earlier so it just returns nan or inf whenever outside a certain range.

**Comment by jzmaddock** [comment]
Apologies, just catching up with this, can someone please open a Boost issue for this with the arguments that cause the issue?

The presence of that assertion being triggered suggests something has gone *very* wrong in the calculation of the leading power terms, so I'd be interested to see what's happening there.

**Comment by Gabriel-Kissin** [comment]
> Apologies, just catching up with this, can someone please open a Boost issue for this with the arguments that cause the issue?

Sure - have opened an issue for this at https://github.com/boostorg/math/issues/1006.  

Suggest that the `beta` part of this issue move there, and the `t` part (`scipy.special.stdtrit` - which is an entirely separate issue) remain here.

**Comment by mborland** [comment]
Sorry for the delay at looking at this as well. I will have time today to work on this one.

**Comment by mdhaber** [comment]
@Gabriel-Kissin, can you build SciPy from source?

If so, could you check out the latest commit in the Boost submodule and see if the issue has been resolved (or at least improved - no crash)? boostorg/math#961 and boostorg/math#962 might have helped here.

**Comment by Gabriel-Kissin** [comment]
Sure, done. Apologies for the delay, I just started a new job so waited till the weekend. Here is what I got, after building a development version by cloning the scipy repo, and running the code in the original issue again:  

## `t` 
no change in output or errors:

```
import scipy.stats
print([shape_param.domain for shape_param in scipy.stats.t._shape_info()])
# [[5e-324, inf]]
print(scipy.stats.t(1e-10).ppf(.2))
# -1e+100
print(scipy.stats.t(1e-20).ppf(.2))
# 1e+100
print(scipy.stats.t(1e-30).ppf(.2))
# KERNEL CRASHES with message: `STOP  SMALL, X, BIG not monotone in INVR`
```


## `beta`
Output has changed for the first example; error message changed for the second example; doesn't crash any more for the third example.

```
import scipy.stats
print([shape_param.domain for shape_param in scipy.stats.beta._shape_info()])
# [[5e-324, inf], [5e-324, inf]]

print(scipy.stats.beta(a=1e-308, b=5).ppf(.2))
# WAS 0.0
# NOW 2.225073858507201e-308

print(scipy.stats.beta(a=1e-309, b=5).ppf(.2))
# WAS OverflowError: Error in function boost::math::tgamma<d>(d): Overflow Error
# (raises error, but kernel doesn't crash)
# NOW OverflowError: Error in function boost::math::beta<d>(%1%,%1%): 
# (raises error, but kernel doesn't crash)

print(scipy.stats.beta(a=1e-323, b=5).ppf(.2))
# KERNEL CRASHES with message: `Assertion failed: *p_derivative >= 0, file ..\..\scipy\_lib\boost/boost/math/special_functions/beta.hpp, line 739`
# NOW OverflowError: Error in function boost::math::beta<d>(%1%,%1%): 
# (raises error, but kernel doesn't crash)
```

**Comment by mdhaber** [comment]
Thanks @Gabriel-Kissin. I think we'll be able to fix the `beta` kernel crash in gh-19174. The `OverflowError` is something I think we'll need to live with for now.

**Comment by ilayn** [comment]
After #19560 merged on main, I'm getting

```python
>> import scipy.stats
>> print(scipy.stats.t(1e-30).ppf(.2))
1.1559163026144151e+118
```

Hence at least the `t` part is resolved in the coming v.1.13.

**Comment by mdhaber** [comment]
This appears to have been fixed at some point. In main, the [Beta example](https://github.com/scipy/scipy/issues/18906#issuecomment-1703961052) no longer raises any errors; it just returns 0.0. The `t` issue was fixed before, as @ilayn noted above. If someone else could check me, I'd appreciate it. In the meantime, I'll go ahead and close.

**Comment by Gabriel-Kissin** [comment]
Thanks @mdhaber and everyone else who has worked on this. I've checked on Windows and Mac using `scipy 1.13.0` and can confirm:

#### 1) The issue with `t` appears to be fixed:  
```
scipy.stats.t(1e-30).ppf(.2)
# 1.1559163026144154e+118
```

\- this previously caused a kernel crash.


#### 2) The issue with `beta` is partially fixed:

Previously, running

```
print(scipy.stats.beta(a=1e-309, b=5).ppf(.2))
print(scipy.stats.beta(a=1e-323, b=5).ppf(.2))
```
would raise an `OverflowError` for the first, and crash the kernel for the second. Whereas now, both lines raise `OverflowError`. This has been discussed above, and as @mdhaber said "The OverflowError is something I think we'll need to live with for now."

Thanks all!

**Comment by mdhaber** [comment]
Actually, with the development version, I'm not even getting the overflow in the development version of SciPy. If you're interested, you can try it out the [nightly wheels](https://anaconda.org/scientific-python-nightly-wheels/scipy).
